### PR TITLE
SAK-41409 Assignments handle submissions that can't be automatically removed

### PR DESCRIPTION
**Pull request by ern**

### CODE REVIEW

**Comment by adrianmticarum** [comment]
Looks good! But we would like to review it deeply tomorrow, please.

**Comment by fostersdesign** [comment]
@adrianmticarum, have you had a chance to review this further?

**Comment by adrianmticarum** [comment]
Hi @fostersdesign! Not yet but we are reviewing this right now. Next week my collegue @jesusmmp will give you an answer!

**Comment by fostersdesign** [comment]
Thanks, @adrianmticarum.

File: assignment/impl/src/java/org/sakaiproject/assignment/impl/persistence/AssignmentRepositoryImpl.java [hunk 1]
```diff
@@ -211,9 +212,22 @@ public AssignmentSubmission findSubmissionForUser(String assignmentId, String us
...
+                        log.error("For assignment {} {} submissions found for user: {}, can only remove {} which is not enough to create a unique submission.", assignmentId, submissions.size(), userId, canRemove.size());
+
+                        // delete those we can remove
+                        submissions.removeAll(canRemove);
                         canRemove.forEach(s -> deleteSubmission(s.getId()));
-                        throw new NonUniqueResultException(sizeDiff);
+
+                        // for the rest we don't actually remove the submission but disassociate it from the submitter
+                        // and use the first submission as the one to return to the ui
+                        submissions.sort(Comparator.comparing(AssignmentSubmission::getDateCreated));
```
Comment (jesusmmp) [inline assignment/impl/src/java/org/sakaiproject/assignment/impl/persistence/AssignmentRepositoryImpl.java]: I think we can remove this sort and in submissions add next restriction.
**.addOrder(Order.desc("dateModified")**

_If the assignment is not a group assignment_
After that, iterate submission until getDateSubmitted() != null. 
If we found one, then return this one. If not found return first one.

Adding findSubmissionForUser method code in the conversation.

**Comment by jesusmmp** [comment]
@Override
	public AssignmentSubmission findSubmissionForUser(String assignmentId, String userId) {
		final List<AssignmentSubmission> lista = sessionFactory.getCurrentSession()
				.createCriteria(AssignmentSubmission.class).add(Restrictions.eq("assignment.id", assignmentId))
				.createAlias("submitters", "s").add(Restrictions.eq("s.submitter", userId))
				.addOrder(Order.desc("dateModified")).list();

		if (lista.isEmpty()) {
			return null;
		}
		else {
			final Assignment a = (Assignment) sessionFactory.getCurrentSession().createCriteria(Assignment.class)
					.add(Restrictions.eq("id", assignmentId)).uniqueResult();

			if (!a.getIsGroup()) {
				for (AssignmentSubmission assignmentSubmission : lista) {
					if (assignmentSubmission.getDateSubmitted() != null) {
						return assignmentSubmission;
					}
				}
			}
			return lista.get(0);
		}
	}

**Comment by ern** [comment]
Going to close this as it is not part of #7345

**Pull request closed without merging**